### PR TITLE
[SD-619] Enables user to remove credit card on checkout payment step

### PR DESCRIPTION
**Pull request by aplegatt**

### CODE REVIEW

File: frontend/app/controllers/spree/checkout_controller.rb [hunk 1]
```diff
@@ -50,6 +50,13 @@ def update
       end
     end
 
+    def remove_credit_card
```
Comment (nnande) [inline frontend/app/controllers/spree/checkout_controller.rb:53]: also we might need to ensure user can remove only CCs that belong to him

File: frontend/app/controllers/spree/checkout_controller.rb [hunk 2]
```diff
@@ -50,6 +50,13 @@ def update
...
+      card = Spree::CreditCard.find(params[:credit_card])
+      card.payments.joins(:order).where("spree_orders.state != 'complete'").each(&:invalidate!)
+      card.update(removed: true)
```
Comment (nnande) [inline frontend/app/controllers/spree/checkout_controller.rb:56]: @aplegatt Does it make sense to wrap the 2 lines above into a DB transaction?

File: frontend/app/views/spree/checkout/_credit_card.html.erb [hunk 3]
```diff
@@ -5,5 +5,6 @@
     <%= credit_card_icon(card.cc_type) %>
     <span>****<%= card.last_digits %>,&nbsp</span>
     <span><%= card.name %></span>
+    <%= link_to 'remove', remove_credit_card_path(card), method: :post, class: 'btn p-0' %>
```
Comment (damianlegawiec) [inline frontend/app/views/spree/checkout/_credit_card.html.erb:8]: @aplegatt let's use translations

File: frontend/config/routes.rb [hunk 4]
```diff
@@ -9,6 +9,7 @@
   patch '/checkout/update/:state', to: 'checkout#update', as: :update_checkout
   get '/checkout/:state', to: 'checkout#edit', as: :checkout_state
   get '/checkout', to: 'checkout#edit', as: :checkout
+  post '/checkout/remove_credit_card/:credit_card', to: 'checkout#remove_credit_card', as: :remove_credit_card
```
Comment (damianlegawiec) [inline frontend/config/routes.rb:12]: @aplegatt let's move that to API controller please - https://github.com/spree/spree/blob/master/api/app/controllers/spree/api/v2/storefront/account/credit_cards_controller.rb - under `destroy` action

File: core/db/migrate/20201007105639_add_removed_to_credit_cards.rb [hunk 5]
```diff
@@ -0,0 +1,5 @@
+class AddRemovedToCreditCards < ActiveRecord::Migration[6.0]
+  def change
+    add_column :spree_credit_cards, :removed, :boolean, default: false unless column_exists?(:spree_credit_cards, :removed)
```
Comment (damianlegawiec) [inline core/db/migrate/20201007105639_add_removed_to_credit_cards.rb:3]: @aplegatt can we just use `deleted_at` for this?

**Review by przemosk — APPROVED** [review]

**Review by szymoniwacz — APPROVED** [review]

File: api/app/controllers/spree/api/v2/storefront/account/credit_cards_controller.rb [hunk 6]
```diff
@@ -14,6 +14,19 @@ def show
               render_serialized_payload { serialize_resource(resource) }
             end
 
+            def destroy
+              card = spree_current_user.credit_cards.find(params[:credit_card_id])
+
+              ActiveRecord::Base.transaction do
```
Comment (damianlegawiec) [inline api/app/controllers/spree/api/v2/storefront/account/credit_cards_controller.rb:20]: @aplegatt please move this into a service - `CreditCards::Destroy`

File: api/app/controllers/spree/api/v2/storefront/account/credit_cards_controller.rb [hunk 7]
```diff
@@ -14,6 +14,11 @@ def show
               render_serialized_payload { serialize_resource(resource) }
             end
 
+            def destroy
+              card = resource.find(params[:credit_card_id])
```
Comment (damianlegawiec) [inline api/app/controllers/spree/api/v2/storefront/account/credit_cards_controller.rb:18]: @aplegatt I believe `resource` will return the credit card object, you should not need to call find here again

File: core/app/services/spree/credit_cards/destroy.rb [hunk 8]
```diff
@@ -0,0 +1,18 @@
+module Spree
+  module CreditCards
+    class Destroy
+      prepend Spree::ServiceModule::Base
+
+      def call(card:)
+        ActiveRecord::Base.transaction do
+          card.payments.
```
Comment (damianlegawiec) [inline core/app/services/spree/credit_cards/destroy.rb:8]: @aplegatt 

`invalidate!` works only on payments in state `checkout`
for `pending` and `processing` states you need to call `void!`

please see: https://github.com/spree/spree/blob/master/core/app/models/spree/payment.rb#L77

**Review by damianlegawiec — CHANGES_REQUESTED** [review]

File: api/app/controllers/spree/api/v2/storefront/account/credit_cards_controller.rb (hunk 7 above)
Comment (aplegatt) [reply api/app/controllers/spree/api/v2/storefront/account/credit_cards_controller.rb:18]: @damianlegawiec Somehow, `resource` is giving me an array of `Spree::CreditCard`s here. I'm going to check the finder maybe something is wrong there.

Comment (damianlegawiec) [reply api/app/controllers/spree/api/v2/storefront/account/credit_cards_controller.rb:18]: @aplegatt yes, the `resource` method is wrong it should be

```ruby
def scope
  resource_finder.new.execute(scope: scope, params: params)
end

def resource
  scope.find(params[:id])
end
```

Comment (damianlegawiec) [reply api/app/controllers/spree/api/v2/storefront/account/credit_cards_controller.rb:18]: @aplegatt 

also the `index` method is wront as well...

please replicate https://github.com/spree/spree/blob/a4efe63593e13d5e3147ba5995e72e008ff609a2/api/app/controllers/spree/api/v2/storefront/account/orders_controller.rb#L6

File: api/app/controllers/spree/api/v2/storefront/account/credit_cards_controller.rb [hunk 9]
```diff
@@ -14,6 +14,10 @@ def show
               render_serialized_payload { serialize_resource(resource) }
             end
 
+            def destroy
+              Spree::Api::Dependencies.storefront_credit_cards_destroy_service.constantize.new.call(card: resource)
```
Comment (szymoniwacz) [inline api/app/controllers/spree/api/v2/storefront/account/credit_cards_controller.rb:18]:
```suggestion
              Spree::Api::Dependencies.storefront_credit_cards_destroy_service.constantize.call(card: resource)
```

File: frontend/app/assets/javascripts/spree/frontend.js [hunk 10]
```diff
@@ -44,6 +44,7 @@ Spree.routes.api_tokens = Spree.pathFor('api_tokens')
 Spree.routes.ensure_cart = Spree.pathFor('ensure_cart')
 Spree.routes.api_v2_storefront_cart_apply_coupon_code = Spree.pathFor('api/v2/storefront/cart/apply_coupon_code')
 Spree.routes.api_v2_storefront_cart_remove_coupon_code = Spree.pathFor('api/v2/storefront/cart/remove_coupon_code')
+Spree.routes.remove_credit_card = function(id) { return Spree.pathFor('api/v2/storefront/account/credit_cards/' + id + '/remove_credit_card') }
```
Comment (szymoniwacz) [inline frontend/app/assets/javascripts/spree/frontend.js:47]:
```suggestion
Spree.routes.api_v2_storefront_remove_credit_card = function(id) { return Spree.pathFor('api/v2/storefront/account/credit_cards/' + id + '/remove_credit_card') }
```

File: api/app/controllers/spree/api/v2/storefront/account/credit_cards_controller.rb (hunk 7 above)
Comment (aplegatt) [reply api/app/controllers/spree/api/v2/storefront/account/credit_cards_controller.rb]: It seems I'm getting a `SystemStackError: stack level too deep` with the `scope` method looking like that due to the recursion. But I was able to make finder work by adding this line: https://github.com/spree/spree/pull/10534/files#diff-69f5940e7246fe4920140f6901ac6242634718f5451586188111b5225ef26d36R7

**Pull request closed without merging**